src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/NetEventSource.WebSockets.cs

```diff
@@ -0,0 +1,286 @@
+
+        [NonEvent]
+        public static void ReceiveAsyncPrivateStarted(object? obj, int bufferLength, [CallerMemberName] string? memberName = null)
+        {
+            Debug.Assert(Log.IsEnabled());
+            Log.ReceiveStart(IdOf(obj), memberName ?? MissingMember, bufferLength);
+        }
+
+        [NonEvent]
+        public static void ReceiveAsyncPrivateCompleted(object? obj, [CallerMemberName] string? memberName = null)
+        {
+            Debug.Assert(Log.IsEnabled());
+            Log.ReceiveStop(IdOf(obj), memberName ?? MissingMember);
+        }
+
+        //
+        // SendFrameAsync
+        //
+
+        [Event(SendStartId, Keywords = Keywords.Debug, Level = EventLevel.Verbose)]
+        private void SendStart(string objName, string memberName, string opcode, int bufferLength) =>
+            WriteEvent(SendStartId, objName, memberName, opcode, bufferLength);
+
+        [Event(SendStopId, Keywords = Keywords.Debug, Level = EventLevel.Verbose)]
+        private void SendStop(string objName, string memberName) =>
+            WriteEvent(SendStopId, objName, memberName);
+
+        [NonEvent]
+        public static void SendFrameAsyncStarted(object? obj, string opcode, int bufferLength, [CallerMemberName] string? memberName = null)
+        {
+            Debug.Assert(Log.IsEnabled());
+            Log.SendStart(IdOf(obj), memberName ?? MissingMember, opcode, bufferLength);
+        }
+
+        [NonEvent]
+        public static void SendFrameAsyncCompleted(object? obj, [CallerMemberName] string? memberName = null)
+        {
+            Debug.Assert(Log.IsEnabled());
+            Log.SendStop(IdOf(obj), memberName ?? MissingMember);
+        }
+
+        //
+        // AsyncMutex
+        //
+
+        [Event(MutexEnterId, Keywords = Keywords.Debug, Level = EventLevel.Verbose)]
+        private void MutexEnter(string objName, string memberName) =>
+            WriteEvent(MutexEnterId, objName, memberName);
+
+        [Event(MutexExitId, Keywords = Keywords.Debug, Level = EventLevel.Verbose)]
+        private void MutexExit(string objName, string memberName) =>
+            WriteEvent(MutexExitId, objName, memberName);
+
+        [Event(MutexContendedId, Keywords = Keywords.Debug, Level = EventLevel.Verbose)]
+        private void MutexContended(string objName, string memberName, int queueLength) =>
+            WriteEvent(MutexContendedId, objName, memberName, queueLength);
+
+        [NonEvent]
+        public static void MutexEntered(object? obj, [CallerMemberName] string? memberName = null)
+        {
+            Debug.Assert(Log.IsEnabled());
+            Log.MutexEnter(IdOf(obj), memberName ?? MissingMember);
+        }
+
+        [NonEvent]
+        public static void MutexExited(object? obj, [CallerMemberName] string? memberName = null)
+        {
+            Debug.Assert(Log.IsEnabled());
+            Log.MutexExit(IdOf(obj), memberName ?? MissingMember);
+        }
```

Comment:
I think repeating at the call site adds more information (i.e. from which function the lock was acquired). Doing the same from within AsyncMutex does not seem trivial to me.

Reply:
> from which function the lock was acquired

exactly, this was the most interesting part 😄

I've tried activities but, I assume, because sometimes lock is taken from one async execution path, but released in some other "observed" background task, activities were not working correctly with them 🥲